File: internal/github/cached.go
package github

import (
	"context"
	"fmt"
	"time"

	"github.com/blakewilliams/gg/internal/cache"
)

// CachedClient wraps a Client with TanStack Query-style caching.
type CachedClient struct {
	client *Client
	cache  *cache.Cache
}

// NewCachedClient creates a CachedClient wrapping the given Client.
func NewCachedClient(client *Client, opts cache.Options) *CachedClient {
	return &CachedClient{
		client: client,
		cache:  cache.New(opts),
	}
}

// GCInterval returns the configured GC interval for tick commands.
func (c *CachedClient) GCInterval() time.Duration {
	return c.cache.GCInterval()
}

// GC runs garbage collection on the cache.
func (c *CachedClient) GC() {
	c.cache.GC()
}

// InvalidatePR removes cached data for a specific PR.
func (c *CachedClient) InvalidatePR(owner, repo string, number int) {
	key := fmt.Sprintf("%s/%s:pull:%d", owner, repo, number)
	c.cache.Invalidate(key)
	c.cache.Invalidate(fmt.Sprintf("%s/%s:pull-files:%d", owner, repo, number))
}

// InvalidateAll clears the entire cache.
func (c *CachedClient) InvalidateAll() {
	c.cache.InvalidatePrefix("")
}

// --- Cached query methods ---

func (c *CachedClient) ListPullRequests(owner, repo string) (data []PullRequest, found bool, refetch func() ([]PullRequest, error)) {
	key := fmt.Sprintf("%s/%s:pulls", owner, repo)
	fetchFn := func() ([]PullRequest, error) {
		return c.client.ListPullRequests(context.Background(), owner, repo)
	}
	data, found, _, refetch = cache.Query(c.cache, key, fetchFn)
	return
}

func (c *CachedClient) GetReviews(owner, repo string, number int) (data []Review, found bool, refetch func() ([]Review, error)) {
	key := fmt.Sprintf("%s/%s:reviews:%d", owner, repo, number)
	fetchFn := func() ([]Review, error) {
		return c.client.GetReviews(context.Background(), owner, repo, number)
	}
	data, found, _, refetch = cache.Query(c.cache, key, fetchFn)
	return
}

func (c *CachedClient) GetIssueComments(owner, repo string, number int) (data []IssueComment, found bool, refetch func() ([]IssueComment, error)) {
	key := fmt.Sprintf("%s/%s:comments:%d", owner, repo, number)
	fetchFn := func() ([]IssueComment, error) {
		return c.client.GetIssueComments(context.Background(), owner, repo, number)
	}
	data, found, _, refetch = cache.Query(c.cache, key, fetchFn)
	return
}

func (c *CachedClient) GetReviewComments(owner, repo string, number int) (data []ReviewComment, found bool, refetch func() ([]ReviewComment, error)) {
	key := fmt.Sprintf("%s/%s:review-comments:%d", owner, repo, number)
	fetchFn := func() ([]ReviewComment, error) {
		return c.client.GetReviewComments(context.Background(), owner, repo, number)
	}
	data, found, _, refetch = cache.Query(c.cache, key, fetchFn)
	return
}

func (c *CachedClient) GetCheckRuns(owner, repo, ref string) (data []CheckRun, found bool, refetch func() ([]CheckRun, error)) {
	key := fmt.Sprintf("%s/%s:check-runs:%s", owner, repo, ref)
	fetchFn := func() ([]CheckRun, error) {
		return c.client.GetCheckRuns(context.Background(), owner, repo, ref)
	}
	data, found, _, refetch = cache.Query(c.cache, key, fetchFn)
	return
}

func (c *CachedClient) GetPullRequestFiles(owner, repo string, number int) (data []PullRequestFile, found bool, refetch func() ([]PullRequestFile, error)) {
	key := fmt.Sprintf("%s/%s:pull-files:%d", owner, repo, number)
	fetchFn := func() ([]PullRequestFile, error) {
		return c.client.GetPullRequestFiles(context.Background(), owner, repo, number)
	}
	data, found, _, refetch = cache.Query(c.cache, key, fetchFn)
	return
}

func (c *CachedClient) GetFileContent(owner, repo, filename, ref string) (data string, found bool, refetch func() (string, error)) {
	key := fmt.Sprintf("%s/%s:file-content:%s:%s", owner, repo, ref, filename)
	fetchFn := func() (string, error) {
		return c.client.GetFileContent(context.Background(), owner, repo, filename, ref)
	}
	data, found, _, refetch = cache.Query(c.cache, key, fetchFn)
	return
}

// FetchFileContent synchronously fetches file content, using the cache.
func (c *CachedClient) FetchFileContent(owner, repo, filename, ref string) (string, error) {
	key := fmt.Sprintf("%s/%s:file-content:%s:%s", owner, repo, ref, filename)
	fetchFn := func() (string, error) {
		return c.client.GetFileContent(context.Background(), owner, repo, filename, ref)
	}

	data, found, _, refetchFn := cache.Query(c.cache, key, fetchFn)
	if found {
		return data, nil
	}
	if refetchFn != nil {
		return refetchFn()
	}
	return "", fmt.Errorf("failed to fetch %s", filename)
}

// --- Direct fetch methods (no caching) ---

func (c *CachedClient) FetchAuthenticatedUser() (User, error) {
	return c.client.GetAuthenticatedUser(context.Background())
}

func (c *CachedClient) FetchInbox(username, nwo string) ([]InboxPR, error) {
	prs, err := c.client.FetchInboxPRs(context.Background(), username, nwo)
	if err != nil {
		return nil, err
	}
	prs, _ = c.client.EnrichPRs(context.Background(), prs, username)
	return prs, nil
}

func (c *CachedClient) FetchPR(owner, repo string, number int) (PullRequest, error) {
	return c.client.GetPullRequest(context.Background(), owner, repo, number)
}

func (c *CachedClient) FetchPRByBranch(owner, repo, branch string) (PullRequest, error) {
	return c.client.GetPullRequestByBranch(context.Background(), owner, repo, branch)
}

func (c *CachedClient) CreateReviewComment(owner, repo string, number int, body, commitID, path string, line int, side string, startLine int, startSide string) (ReviewComment, error) {
	comment, err := c.client.CreateReviewComment(context.Background(), owner, repo, number, body, commitID, path, line, side, startLine, startSide)
	if err != nil {
		return ReviewComment{}, err
	}
	c.cache.Invalidate(fmt.Sprintf("%s/%s:review-comments:%d", owner, repo, number))
	return comment, nil
}

func (c *CachedClient) ReplyToReviewComment(owner, repo string, number int, commentID int, body string) (ReviewComment, error) {
	comment, err := c.client.ReplyToReviewComment(context.Background(), owner, repo, number, commentID, body)
	if err != nil {
		return ReviewComment{}, err
	}
	c.cache.Invalidate(fmt.Sprintf("%s/%s:review-comments:%d", owner, repo, number))
	return comment, nil
}

File: rust/src/github/cached.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};

use tokio::sync::RwLock;

use super::Client;
use super::types::*;

const GC_INTERVAL: Duration = Duration::from_secs(60);
const STALE_TIME: Duration = Duration::from_secs(30);
const GC_TIME: Duration = Duration::from_secs(300);

struct CacheEntry<T> {
    data: T,
    fetched_at: Instant,
    last_accessed: Instant,
}

pub struct CachedClient {
    client: Client,
    pr_list: Arc<RwLock<HashMap<String, CacheEntry<Vec<PullRequest>>>>>,
    reviews: Arc<RwLock<HashMap<String, CacheEntry<Vec<Review>>>>>,
    review_comments: Arc<RwLock<HashMap<String, CacheEntry<Vec<ReviewComment>>>>>,
    issue_comments: Arc<RwLock<HashMap<String, CacheEntry<Vec<IssueComment>>>>>,
    pr_files: Arc<RwLock<HashMap<String, CacheEntry<Vec<PullRequestFile>>>>>,
    check_runs: Arc<RwLock<HashMap<String, CacheEntry<CheckRunsResponse>>>>,
}

impl CachedClient {
    pub fn new(client: Client) -> Self {
        Self {
            client,
            pr_list: Arc::new(RwLock::new(HashMap::new())),
            reviews: Arc::new(RwLock::new(HashMap::new())),
            review_comments: Arc::new(RwLock::new(HashMap::new())),
            issue_comments: Arc::new(RwLock::new(HashMap::new())),
            pr_files: Arc::new(RwLock::new(HashMap::new())),
            check_runs: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn gc_interval(&self) -> Duration {
        GC_INTERVAL
    }

    pub async fn gc(&self) {
        let now = Instant::now();
        Self::gc_map(&self.pr_list, now, GC_TIME).await;
        Self::gc_map(&self.reviews, now, GC_TIME).await;
        Self::gc_map(&self.review_comments, now, GC_TIME).await;
        Self::gc_map(&self.issue_comments, now, GC_TIME).await;
        Self::gc_map(&self.pr_files, now, GC_TIME).await;
        Self::gc_map(&self.check_runs, now, GC_TIME).await;
    }

    async fn gc_map<T>(map: &Arc<RwLock<HashMap<String, CacheEntry<T>>>>, now: Instant, max_age: Duration) {
        let mut m = map.write().await;
        m.retain(|_, v| now.duration_since(v.last_accessed) < max_age);
    }

    pub async fn invalidate_all(&self) {
        self.pr_list.write().await.clear();
        self.reviews.write().await.clear();
        self.review_comments.write().await.clear();
        self.issue_comments.write().await.clear();
        self.pr_files.write().await.clear();
        self.check_runs.write().await.clear();
    }

    pub async fn authenticated_user(&self) -> anyhow::Result<User> {
        self.client.authenticated_user().await
    }

    pub async fn pull_requests(
        &self,
        owner: &str,
        repo: &str,
    ) -> anyhow::Result<Vec<PullRequest>> {
        let key = format!("{owner}/{repo}/pulls");
        if let Some(cached) = self.get_cached(&self.pr_list, &key).await {
            return Ok(cached);
        }
        let data = self.client.pull_requests(owner, repo).await?;
        self.set_cached(&self.pr_list, key, data.clone()).await;
        Ok(data)
    }

    pub async fn reviews(
        &self,
        owner: &str,
        repo: &str,
        number: u64,
    ) -> anyhow::Result<Vec<Review>> {
        let key = format!("{owner}/{repo}/pulls/{number}/reviews");
        if let Some(cached) = self.get_cached(&self.reviews, &key).await {
            return Ok(cached);
        }
        let data = self.client.reviews(owner, repo, number).await?;
        self.set_cached(&self.reviews, key, data.clone()).await;
        Ok(data)
    }

    pub async fn review_comments(
        &self,
        owner: &str,
        repo: &str,
        number: u64,
    ) -> anyhow::Result<Vec<ReviewComment>> {
        let key = format!("{owner}/{repo}/pulls/{number}/comments");
        if let Some(cached) = self.get_cached(&self.review_comments, &key).await {
            return Ok(cached);
        }
        let data = self.client.review_comments(owner, repo, number).await?;
        self.set_cached(&self.review_comments, key, data.clone()).await;
        Ok(data)
    }

    pub async fn issue_comments(
        &self,
        owner: &str,
        repo: &str,
        number: u64,
    ) -> anyhow::Result<Vec<IssueComment>> {
        let key = format!("{owner}/{repo}/issues/{number}/comments");
        if let Some(cached) = self.get_cached(&self.issue_comments, &key).await {
            return Ok(cached);
        }
        let data = self.client.issue_comments(owner, repo, number).await?;
        self.set_cached(&self.issue_comments, key, data.clone()).await;
        Ok(data)
    }

    pub async fn pr_files(
        &self,
        owner: &str,
        repo: &str,
        number: u64,
    ) -> anyhow::Result<Vec<PullRequestFile>> {
        let key = format!("{owner}/{repo}/pulls/{number}/files");
        if let Some(cached) = self.get_cached(&self.pr_files, &key).await {
            return Ok(cached);
        }
        let data = self.client.pr_files(owner, repo, number).await?;
        self.set_cached(&self.pr_files, key, data.clone()).await;
        Ok(data)
    }

    pub async fn check_runs(
        &self,
        owner: &str,
        repo: &str,
        git_ref: &str,
    ) -> anyhow::Result<CheckRunsResponse> {
        let key = format!("{owner}/{repo}/commits/{git_ref}/check-runs");
        if let Some(cached) = self.get_cached(&self.check_runs, &key).await {
            return Ok(cached);
        }
        let data = self.client.check_runs(owner, repo, git_ref).await?;
        self.set_cached(&self.check_runs, key, data.clone()).await;
        Ok(data)
    }

    pub async fn pull_request(
        &self,
        owner: &str,
        repo: &str,
        number: u64,
    ) -> anyhow::Result<PullRequest> {
        self.client.pull_request(owner, repo, number).await
    }

    pub async fn pull_request_by_branch(
        &self,
        owner: &str,
        repo: &str,
        branch: &str,
    ) -> anyhow::Result<Option<PullRequest>> {
        self.client.pull_request_by_branch(owner, repo, branch).await
    }

    pub async fn create_review_comment(
        &self,
        owner: &str,
        repo: &str,
        number: u64,
        body: &str,
        commit_id: &str,
        path: &str,
        line: i32,
        side: &str,
    ) -> anyhow::Result<ReviewComment> {
        let comment = self
            .client
            .create_review_comment(owner, repo, number, body, commit_id, path, line, side)
            .await?;
        // Invalidate review comments cache
        let key = format!("{owner}/{repo}/pulls/{number}/comments");
        self.review_comments.write().await.remove(&key);
        Ok(comment)
    }

    pub async fn reply_to_comment(
        &self,
        owner: &str,
        repo: &str,
        number: u64,
        comment_id: u64,
        body: &str,
    ) -> anyhow::Result<ReviewComment> {
        let comment = self
            .client
            .reply_to_comment(owner, repo, number, comment_id, body)
            .await?;
        let key = format!("{owner}/{repo}/pulls/{number}/comments");
        self.review_comments.write().await.remove(&key);
        Ok(comment)
    }

    async fn get_cached<T: Clone>(
        &self,
        map: &Arc<RwLock<HashMap<String, CacheEntry<T>>>>,
        key: &str,
    ) -> Option<T> {
        let mut m = map.write().await;
        if let Some(entry) = m.get_mut(key) {
            if entry.fetched_at.elapsed() < STALE_TIME {
                entry.last_accessed = Instant::now();
                return Some(entry.data.clone());
            }
        }
        None
    }

    async fn set_cached<T>(
        &self,
        map: &Arc<RwLock<HashMap<String, CacheEntry<T>>>>,
        key: String,
        data: T,
    ) {
        let now = Instant::now();
        map.write().await.insert(
            key,
            CacheEntry {
                data,
                fetched_at: now,
                last_accessed: now,
            },
        );
    }
}
